### PR TITLE
[stable/grafana] Support Provisioner Configuration

### PR DESCRIPTION
**Pull request by BugRoger**
This PR extends the `stable/grafana` chart to allow for outside configuration of provisioners. 

With Grafana 5 it will be possible to manage datasources and dashboards using provisioners.  Adding a provisioner is done by adding one or more yaml configuration files in the `provisioning/datasources` and `provisioning/dashboards` directory. See also: http://docs.grafana.org/administration/provisioning/

For datasources the file can contain a list of datasources that will be added or updated during start-up. If the datasource already exists, Grafana will update it to match the configuration file. 

For dashboards the files contain the configuration for provisioning from external resources. It currently only support reading dashboards from file but more provisioners are in the works.

This functionality partly clashes with the jobs provided by this Helm chart. In order to support both Grafana 4/5 in the same chart, I think, it is a sensible approach to allow for both provisioning mechanisms. In the absence of more sophisticated provisioners (that allow the decoupling from Helm and dashboards) it is anyway required to load the dashboard from the filesystem and inject it using the existing mechanism.

### CODE REVIEW

**Comment by NicolasT** [comment]
Looking forward to this to land. Too bad though that creating a default datasource for e.g. Prometheus deployed by the stable Chart (in a Chart having both Grafana and Prometheus as requirements) still won't be possible using this approach, because its service name/DNS entry depends on the Helm release name 😞

**Comment by monotek** [comment]
@BugRoger 
Looking forward to see this merged. Would you mind to fix the the merge conflict?

**Comment by BugRoger** [comment]
Fixed. It was just some other change increasing the version

**Comment by pesimon** [comment]
Can confirm that it works like a charm and grafana 5.0.0 was just released
https://github.com/grafana/grafana/releases

File: stable/grafana/templates/deployment.yaml
Comment (unguiculus) [inline stable/grafana/templates/deployment.yaml:100]: Don't use the `default` function. Defaults come from `values.yaml` anyways which is better. Apply to all your changes.

**Review by unguiculus — CHANGES_REQUESTED** [review]

**Comment by jpds** [comment]
This merge proposal doesn't actually change the default version to `5.0.0` (relevant issue: #3937).

**Comment by sstarcher** [comment]
This should bump the version to 5.0.1 and add the AppVersion to the charts.yaml

File: stable/grafana/templates/deployment.yaml
Comment (unguiculus) [reply stable/grafana/templates/deployment.yaml:100]: I figured this is common practice in the chart. So, let's leave it for now. Would be cool to get that fixed in a separate PR.

**Review by unguiculus — CHANGES_REQUESTED** [review]

**Review by unguiculus — APPROVED** [review]

**Comment by sstarcher** [comment]
@BugRoger could you fix the linting issues from the circleci build

**Comment by richerve** [comment]
@BugRoger are you still working on this, if not I can help

**Comment by richerve** [comment]
In my opinion is better to bump the minor version and remove at least the support to configure datasources using the Job, it is a hack that is difficult to maintain and not needed anymore. In general the dashboards Job will also not be needed but I think is handy to have the functionality to import remote dashboards.

About bumping the minor version (or major) of the Chart, Grafana is not a project that has Long Term Support releases so keeping a Chart that is compatible with version 4.6.x and another compatible with >= 5 is not a big issue I think, eventually the Chart versions compatible with Grafana <= 4 will not be updated anymore and this will happen sooner than later.

I want to use this Chart on the *kube_prometheus* "meta Chart" in coreos/prometheus-operator#1127, they are maintaining what is basically a copy of the stable Chart but adding *grafana-watcher* that essentially serves the same purpose as the Jobs here, both cases are hacks that with Grafana >= 5 are not required.

**Comment by BugRoger** [comment]
Fixed the linter nitpicks. Bumped the version. Rebased on master.

**Comment by richerve** [comment]
Any way to move this forward?

**Comment by monotek** [comment]
@BugRoger 

Please update to the newest Grafana 5.0.4 Docker image as the older images don't work with Kubernetes 1.7.14, 1.8.9, or 1.9.4. This would probably also fix the build. 

See: https://github.com/grafana/grafana-docker/issues/140

**Comment by BugRoger** [comment]
Rebased. Updated Image to 5.0.4. Added `appVersion` to make Linter happy. Tests are green.

**Comment by richerve** [comment]
@BugRoger, sadly again conflicting files. I think this should be coordinated with someone that can merge, do the fix and immediately apply it. I'm waiting for this PR to be merged to work on the integration in https://github.com/coreos/kube-prometheus

**Comment by knopkodav** [comment]
The same issue on 1.9.6-gke.0

**Comment by matthope** [comment]
How does this compare to #4541  ?

**Comment by richerve** [comment]
@matthope In concept both are the same in approach are different, this PR include changes to support the new Grafana >= 5 config but also maintains backward compatibility. The PR on #4541 removes all backwards compatibility.

@BugRoger friendly ping about updating the branch.

**Comment by rtluckie** [comment]
please check out #4713 before merging.

**Comment by BugRoger** [comment]
It seems i can't fix this. There's a deadlock between the different checks.

The linter requires me to set the version to version in `chart.yaml` to `0.8.5`. This causes a conflict against master. That fails the e2e tests. 

I can either resolve the conflict or fix the version.

**Comment by sstarcher** [comment]
master is already at 0.8.5.  You need to rebase and bump it to 0.8.6 or 0.9.0

**Comment by BugRoger** [comment]
Ah, rebased and bumped the version again....

**Comment by unguiculus** [comment]
I'm inclined to close this in favor of #4713.

**Comment by sstarcher** [comment]
@unguiculus if #4713 is merged it would make this PR no longer needed since #4713 is such a large overhaul and simplification.

I have not tested #4713, but the changes themselves are heading in a good direction.

**Comment by unguiculus** [comment]
Closing in favor of #4713. Thanks @BugRoger.

**Pull request closed without merging**
